### ISSUE TITLE
Performance: Font caching redundantly executing Font.size()

### ISSUE THREAD

**Issue by nrichards** [IndiePython/nodezator#50]
- Seen every time
- Environment: Nodezator 1.3.13, M1 Macbook Pro OS 13.4, Python 3.11

# Issue

Launching the app results in 3,864 calls to Font.size(" "), 24 of which are unique. According to [snakeviz](https://jiffyclub.github.io/snakeviz/) ([usage](https://stackoverflow.com/a/71980877/915990)) this takes 0.99 seconds on my computer. 

# Expectation

I'd expect Font.size(" ") to be called 24 times, as there is a caching system in place.

# Reproduction

Count redundant executions

1. Add to fontsman/cache.py before the 1st font.size(" "):
   1. `print(f"A-{desired_height}: {font.name}")`
1. Add before the 2nd font.size(" "), inside the while 
   1. `print(f"B-{size}: {font.name}")`
1. run the program, see logs
2. count all the calls to font.size, trimming any spurious leading/trailing logs (on Mac: copy log into pasteboard CMD+C, open terminal, run `pbpaste | wc -l`)
3. count the unique calls to font.size (on Mac: copy log into pasteboard CMD+C, open terminal, run `pbpaste | sort -u | wc -l`)

# Profiling

```sh
$ pip install snakeviz
$ python3 -m cProfile -o program.prof -m nodezator
$ snakeviz program.prof
```

# Resources

[font_size_for_launch.txt](https://github.com/IndiePython/nodezator/files/11641790/font_size_for_launch.txt)

<img width="1229" alt="image" src="https://github.com/IndiePython/nodezator/assets/704698/4bc9e1ff-b873-4fa7-b1ef-1025acc0077f">

**Comment by KennedyRichard** [comment]
Mr. Richards,

First of all, thank you for taking the time to help improve Nodezator. This detailed report is much appreciated and shows your dedication. I'm just stopping by to acknowledge receipt/awareness of this issue and to let you know I'll give you a proper reply in a couple of days (excluding Sunday).

Have a nice weekend,

Peace

**Comment by nrichards** [comment]
Thanks for the acknowledgement, I'm glad this appears helpful.

> dedication

I seek to clearly identify areas I find worthy of attention, from my current perspective as a user.

**Comment by nrichards** [comment]
@KennedyRichard Please have a look at referenced experimental fix for font performance, and also for unlikely infinite loop and a crash with too small/large heights.

**Comment by nrichards** [comment]
Note to self: 

I might want to add in another test, for font size of 23. Size of 22 and font size is 24 both return a font and font size of 23 returns a unreachable font size exception. It would be good probably to test for that to validate that it works the way it's supposed to work.

**Comment by KennedyRichard** [comment]
Mr. Richards,

I apologize for my delay in tending to this. I've been working in a new feature for the next release since Monday and only today I was able to reach a point where I could show something  (I posted about it in the discussion regarding current work #34 ).

I'll tend to this issue and the related pull request from tomorrow onwards (probably after lunch, since I have a commitment in the morning).

Peace
